azurelinuxagent/common/event.py

```diff
@@ -393,52 +529,106 @@ def _clean_up_message(message):
+        event.parameters = trimmed_params
+
+    def collect_events(self):
+        """
+        Retuns a list of events that need to sent to the telemetry pipeline and deletes the corresponding files
+        from the events directory.
+        """
+        event_list = TelemetryEventList()
+        event_directory_full_path = os.path.join(conf.get_lib_dir(), EVENTS_DIRECTORY)
+        event_files = os.listdir(event_directory_full_path)
 
-        for param in default_parameters:
-            if param.name not in param_names or set_values_for_agent:
-                # If set_values_for_agent, we disregard any values already set for an existing default property and
-                # replaces it with a latest entry.
-                param_names[param.name] = param.value
+        for event_file in event_files:
+            try:
+                match = EVENT_FILE_REGEX.search(event_file)
+                if match is None:
+                    continue
 
-        parameters = DataContractList(TelemetryEventParam)
-        for name, value in param_names.items():
-            parameters.append(TelemetryEventParam(name, value))
+                event_file_path = os.path.join(event_directory_full_path, event_file)
 
-        return parameters
+                try:
+                    logger.verbose("Processing event file: {0}", event_file_path)
+
+                    with open(event_file_path, "rb") as fd:
+                        event_data = fd.read().decode("utf-8")
+
+                    event = parse_event(event_data)
+
+                    is_legacy_event = match.group('agent_event') is None
+
+                    if is_legacy_event:
+                        # We'll use the file creation time for the event's timestamp
+                        event_file_creation_time_epoch = os.path.getmtime(event_file_path)
+                        event_file_creation_time = datetime.fromtimestamp(event_file_creation_time_epoch)
+
+                        if event.is_extension_event():
+                            EventLogger._trim_extension_event_parameters(event)
+                            self._add_common_event_parameters(event, event_file_creation_time)
+                        else:
+                            self._update_old_event_schema(event, event_file_creation_time)
+
+                    event_list.events.append(event)
+                finally:
+                    os.remove(event_file_path)
+            except Exception as e:
+                logger.warn("Failed to process event file {0}: {1}", event_file, ustr(e))
+
+        return event_list
+
+    def _update_old_event_schema(self, event, event_creation_time):
+        # Ensure that if an agent event is missing a field from the schema defined since 2.2.47, the missing fields
+        # will be appended, ensuring the event schema is complete before the event is reported.
+        new_event = TelemetryEvent()
```

Comment:
This comes from #1757. I discussed this @pgombar - we should not just add any missing fields since we may hide bugs elsewhere or populate with invalid values. Some missing fields must be added, though. For example, OpCode (timestamp) will be in 2.2.46, but not on previous agents.